crates/router/src/core/payment_methods/network_tokenization.rs

```diff
@@ -0,0 +1,700 @@
+
+    let res = response
+        .map_err(|error| {
+            error.change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed)
+        })
+        .attach_printable("Error while receiving response")
+        .and_then(|inner| match inner {
+            Err(err_res) => {
+                let parsed_error: NetworkTokenErrorResponse = err_res
+                    .response
+                    .parse_struct("Card Network Tokenization Response")
+                    .change_context(
+                        errors::NetworkTokenizationError::ResponseDeserializationFailed,
+                    )?;
+                logger::error!(
+                    error_code = %parsed_error.error_info.code,
+                    developer_message = %parsed_error.error_info.developer_message,
+                    "Network tokenization error: {}",
+                    parsed_error.error_message
+                );
+                Err(errors::NetworkTokenizationError::ResponseDeserializationFailed)
+                    .attach_printable(format!("Response Deserialization Failed: {err_res:?}"))
+            }
+            Ok(res) => Ok(res),
+        })
+        .inspect_err(|err| {
+            logger::error!("Error while deserializing response: {:?}", err);
+        })?;
+
+    let network_response: CardNetworkTokenResponse = res
+        .response
+        .parse_struct("Card Network Tokenization Response")
+        .change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed)?;
+
+    let dec_key = tokenization_service.private_key.peek().clone();
+
+    let card_network_token_response = services::decrypt_jwe(
+        &network_response.payload,
+        services::KeyIdCheck::SkipKeyIdCheck,
+        dec_key,
+        jwe::RSA_OAEP_256,
+    )
+    .await
+    .change_context(errors::NetworkTokenizationError::SaveNetworkTokenFailed)
+    .attach_printable(
+        "Failed to decrypt the tokenization response from the tokenization service",
+    )?;
+
+    let cn_response: CardNetworkTokenResponsePayload =
+        serde_json::from_str(&card_network_token_response)
+            .change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed)?;
+    Ok((cn_response.clone(), Some(cn_response.card_reference)))
+}
+
+pub async fn make_card_network_tokenization_request(
+    state: &routes::SessionState,
+    card: &domain::Card,
+    customer_id: &id_type::CustomerId,
+    amount: Option<i64>,
+    currency: Option<storage_enums::Currency>,
+) -> CustomResult<(CardNetworkTokenResponsePayload, Option<String>), errors::NetworkTokenizationError>
+{
+    let card_data = CardData {
+        card_number: card.card_number.clone(),
+        exp_month: card.card_exp_month.clone(),
+        exp_year: card.card_exp_year.clone(),
+        card_security_code: card.card_cvc.clone(),
+    };
+
+    let payload = card_data
+        .encode_to_string_of_json()
+        .and_then(|x| x.encode_to_string_of_json())
```

Comment:
Is the double JSON encoding intentional?

Reply:
yes